feat: Relocate experimental features settings

dcalhoun opened the pull request:
Relocate experimental features settings to reside next to debug settings.

Relates to https://github.com/wordpress-mobile/WordPress-Android/pull/21459#issuecomment-2474767835.

| Light | Dark |
| - | - |
| ![Screenshot_20241114_162233_Jetpack Pre-Alpha](https://github.com/user-attachments/assets/be7101d4-6584-4281-85c9-7cefc116a7bf) | ![Screenshot_20241114_162241_Jetpack Pre-Alpha](https://github.com/user-attachments/assets/08fdc076-33aa-4562-ba52-575392ed1bd6) |
| ![Screenshot_20241114_163559_Jetpack Pre-Alpha](https://github.com/user-attachments/assets/4434b27a-13a1-4d8d-87e7-abb063a11361) | ![Screenshot_20241114_164300_Jetpack Pre-Alpha](https://github.com/user-attachments/assets/776756de-bea0-47c0-84fb-8ef1fc411a14) |

## Testing Instructions

Re-test https://github.com/wordpress-mobile/WordPress-Android/pull/21459 in the new location.

<!-- Test instructions per dependency update: https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/docs/test_instructions_per_dependency_update.md -->

## Regression Notes

1. Potential unintended areas of impact
    Unlikely to impact anything given the narrow focus.
2. What I did to test those areas of impact (or what existing automated tests I relied on)
    N/A.
3. What automated tests I added (or what prevented me from doing so)
    Deemed unnecessary for the experimental features.

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Testing Checklist (strike-out the not-applying and unnecessary ones):

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

dcalhoun committed:
feat: Relocate experimental feature settings

Position experimental features alongside debug settings.

dcalhoun committed:
fix: Titlecase debug settings

Improve consistency of Me screen buttons.

dcalhoun committed:
fix: Remove duplicative key prefix

The prefix is appended by get/set utility methods.

dcalhoun committed:
refactor: Remove unnecessary list abstraction

dcalhoun committed:
refactor: Rename for clarity

dcalhoun committed:
refactor: Improve consistent shorthand usage

dcalhoun committed:
feat: Improve Material 3 styling

dcalhoun committed:
fix: Avoid disabling unchecked switches

dcalhoun committed:
task: Remove previous experimental features app setting

Replaced by experimental features within the Me screen.

dcalhoun committed:
feat: Update experimental feature button icon

Better communicate experimental nature.

dangermattic commented:
<!--
  0 Errors
  3 Warnings: <code>strings.xml</code> files...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>strings.xml</code> files should only be updated on release branches, when the translations are downloaded by our automation.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dcalhoun commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -962,8 +959,11 @@
+
     <!-- Debug settings -->
-    <string name="preference_open_debug_settings" translatable="false">Debug settings</string>
+    <string name="preference_open_debug_settings" translatable="false">Debug Settings</string>
```
Changed to titlecase to match all other buttons/screens.

nbradbury replied:
This is fine since we use titlecase elsewhere, but note that on Android [the recommendation](https://developers.google.com/style/capitalization) is to avoid unnecessary capitalization. This is something we're very inconsistent about.

dcalhoun replied:
Agreed, there is opportunity to improve consistency across the product.

dcalhoun reviewed (commented):
This is my first time using Jetpack Compose, please feel free to provide feedback any aspect. 🙇🏻‍♂️

dcalhoun committed:
style: Address lint errors

dcalhoun committed:
fix: Add NonNull decorator

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/main/MeFragment.kt:
```diff
@@ -216,6 +216,10 @@ class MeFragment : Fragment(R.layout.me_fragment), OnScrollToTopListener {
         if (BuildConfig.IS_JETPACK_APP) meAboutIcon.setImageResource(R.drawable.ic_jetpack_logo_white_24dp)
 
+        rowExperimentalFeaturesSettings.setOnClickListener {
+            ActivityLauncher.viewExperimentalFeatures(activity)
```
The app is failing to build because `activity` is nullable. I recommend using `requireActivity()` instead.

dcalhoun replied:
Addressed in 637e41e0770ec926e35b935faf3326d2e8476c49 . 

The use of `activity` was oversight that came to light when I addressed a nullable lint warning in 4ac3e191248840a9b17b9f165cf83eee245f9ca9. This was intended to be `context`.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/ExperimentalFeaturesActivity.kt:
```diff
@@ -0,0 +1,188 @@
+        Spacer(modifier = Modifier.width(Margin.ExtraLarge.value))
+        Text(
+            text = label,
+            style = MaterialTheme.typography.titleMedium,
```
Instead of `titleMedium` I recommend using `bodyMedium` here.

dcalhoun replied:
Thanks! Addressed in 1a324f0653e32ca89865541c737ec9dcea06c354.

dcalhoun committed:
fix: Pass context instead of activity to avoid crash

The use of `activity` was an oversight. Also, a `@NonNull` decorator was
added to the caller, which led to the enforcement and crash.

dcalhoun committed:
fix: Improve toggle text styles

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/ExperimentalFeaturesActivity.kt:
```diff
@@ -0,0 +1,188 @@
+            }.toMap()
+        }
+
+        CompositionLocalProvider(LocalContext provides LocalContext.current) {
```
I believe `CompositionLocalProvider` isn't necessary here.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/ExperimentalFeaturesActivity.kt:
```diff
@@ -0,0 +1,188 @@
+    Row(
+        verticalAlignment = Alignment.CenterVertically,
+        modifier = modifier
+            .clickable() { onChange(key, label, !enabled) }
```
Minor nit: we can drop the `()` after `clickable` here.

dcalhoun committed:
refactor: Remove unnecessary local provider

This appears no longer necessary for the Jetpack Compose preview after
refactoring of the component state.

dcalhoun committed:
style: Remove unnecessary syntax

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/ExperimentalFeaturesActivity.kt:
```diff
@@ -0,0 +1,184 @@
+import org.wordpress.android.util.extensions.setContent
+
+val experimentalFeatures = listOf(
+    Feature(key = "experimental_block_editor", label = "Experimental block editor"),
```
The labels are hard-coded here so they won't be localized. Is that intentional?

dcalhoun replied:
No, how would you suggest structuring this so that translated strings are possible?

dcalhoun replied:
More specifically: given `stringResource` can only be called within the `Composable`, it cannot be placed here in this `experimentalFeatures` list, how might we leverage translated strings? 

I suppose removing `label` from `Feature` altogether is the best approach, where the label's translated string is dynamically retrieved in the `Composable` via the `key`? 🤔

dcalhoun replied:
Addressed this in fb339a850988afbace05d137a77740d3081f4a3c.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/ExperimentalFeaturesActivity.kt:
```diff
@@ -0,0 +1,184 @@
+@Composable
+fun ExperimentalFeaturesScreen(
+    features: Map<String, Feature>,
+    modifier: Modifier = Modifier,
```
We don't pass a `Modifier` to this function, so I believe this parameter is unnecessary. We can likewise remove the parameter from `FeatureToggle`.

dcalhoun committed:
refactor: Remove unused modifier parameter

nbradbury commented:
@dcalhoun Nice job on your first Compose task! One recommendation I have for future work is to avoid putting the ViewModel and the composables inside the activity. 

I know there are several places we do that in the app, but it's better to have both the ViewModel and the composables as separate files.

The reason for this is to avoid making the mistake of having the ViewModel "aware" of the activity, and likewise for the composables. That mistake was not made here, but it has been made elsewhere in the app.

dcalhoun committed:
feat: Add translated experimental features strings

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/ExperimentalFeaturesActivity.kt:
```diff
@@ -132,17 +139,23 @@ fun ExperimentalFeaturesScreen(
     }
 }
 
+@SuppressLint("DiscouragedApi")
```
Rather than use a discouraged API here, how about changing the `Feature` class to have the actual label string resource ID?

```
data class Feature(
    val enabled: Boolean = false,
    val key: String,
    @StringRes val label: Int
)
```

dcalhoun replied:
One benefit of dynamically retrieving the label based on the `key` is that we no longer have to pass the `label` around when mutating the `enabled` state. The discouraged API appears to relate to performance, but I'm not sure it is very relevant for this specific use case. WDYT?

nbradbury replied:
We're probably fine to stick with this, I just figured since we previously passed around `label` as a string it wouldn't be a problem as an Int.

nbradbury reviewed (approved):
@dcalhoun I'm going to approve this, but there's a failing test in `AnalyticsTrackerNosaraTest`. Once that's taken care of feel free to merge :shipit:

dcalhoun committed:
fix: Align with analytic event naming practices
